readme.md
## ORIGINAL INPUT

```
Building for debugging...
[13/13] Linking SwiftFooPackageTests.xctest
Build complete! (1.78s)
Test Suite 'All tests' started at 2022-12-17 23:08:03.162
Test Suite 'debug.xctest' started at 2022-12-17 23:08:03.163
Test Suite 'QuickSpec' started at 2022-12-17 23:08:03.163
Test Case 'QuickSpec.SwiftFoo, Util, getInput, reads data from file' started at 2022-12-17 23:08:03.163
Test Case 'QuickSpec.SwiftFoo, Util, getInput, reads data from file' passed (0.001 seconds)
Test Case 'QuickSpec.SwiftFoo, Util, getInput, applies the function to the file contents if specified' started at 2022-12-17 23:08:03.164
Test Case 'QuickSpec.SwiftFoo, Util, getInput, applies the function to the file contents if specified' passed (0.0 seconds)
Test Case 'QuickSpec.SwiftFoo, Day1, part1, finds the total of the elve carrying the most calories' started at 2022-12-17 23:08:03.164
Test Case 'QuickSpec.SwiftFoo, Day1, part1, finds the total of the elve carrying the most calories' passed (0.0 seconds)
Test Case 'QuickSpec.SwiftFoo, Day1, part2, finds the sum of the top 3 elves' started at 2022-12-17 23:08:03.164
Test Case 'QuickSpec.SwiftFoo, Day1, part2, finds the sum of the top 3 elves' passed (0.0 seconds)
Test Suite 'QuickSpec' passed at 2022-12-17 23:08:03.164
         Executed 4 tests, with 0 failures (0 unexpected) in 0.001 (0.001) seconds
Test Suite 'debug.xctest' passed at 2022-12-17 23:08:03.164
         Executed 4 tests, with 0 failures (0 unexpected) in 0.001 (0.001) seconds
Test Suite 'All tests' passed at 2022-12-17 23:08:03.164
         Executed 4 tests, with 0 failures (0 unexpected) in 0.001 (0.001) seconds
```

## DESIRED OUTPUT

```
QuickSpec
  SwiftFoo
    Util
      getInput
        reads data from file [✔]
        applies the function to the file contents if specified [✔]
    Day1
      part1
        finds the total of the elf carrying the most calories [✔]
      part2
        finds the sum of the top 3 elves [✔]
```

```
data TestOutput a = Describe String [TestOutput a] 
                  | It String a
    deriving (Eq, Show, Functor, Foldable, Traversable)

data TestResult = Sucess | Failure String
    deriving  (Eq, Show)

type QuickTest = TestOutput TestResult
```


CHANGELOG.md
# Revision history for hurtig

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Hurtig/FooSpec.hs
module Hurtig.FooSpec (spec) where

import Hurtig.Foo
import Test.Hspec

spec :: Spec
spec = do
  describe "Hurtig" $ do
    describe "nothing" $ do
      it "should just succeed" $ do
        foo `shouldBe` 1


test/Hurtig/TestTreeSpec.hs
module Hurtig.TestTreeSpec (spec) where

import Hurtig.QuickLog
import Hurtig.TestTree
import Test.Hspec

spec :: Spec
spec = do
  describe "TestTree" $ do
    describe "adds test cases to an empty tree" $ do
      it "adds a 'Foo/does foo' to an empty tree" $
        insert (QuickLogCasePass ["Foo"] "does foo") (TestTree "Foo" [])
          `shouldBe` TestTree
            "Foo"
            [ TestSpec (TestInfo True "does foo")
            ]
      it "adds a 'Foo,Bar,Baz/does foo' to an empty tree" $
        insert (QuickLogCasePass ["Foo", "Bar", "Baz"] "does foo") (TestTree "Foo" [])
          `shouldBe` TestTree
            "Foo"
            [ TestTree
                "Bar"
                [ TestTree
                    "Baz"
                    [TestSpec (TestInfo True "does foo")]
                ]
            ]
      it "adds a 'Foo,Bar,Baz/does foo' to tree containing Foo,Bar,Baz" $
        insert (QuickLogCasePass ["Foo", "Bar", "Baz"] "does foo") (TestTree "Foo" [TestTree "Bar" [TestTree "Baz" []]])
          `shouldBe` TestTree
            "Foo"
            [ TestTree
                "Bar"
                [ TestTree
                    "Baz"
                    [TestSpec (TestInfo True "does foo")]
                ]
            ]
      it "adds a 'Foo,Bar,Baz/does foo' to tree containing Foo,(Bar/does bar),Baz" $
        insert (QuickLogCasePass ["Foo", "Bar", "Baz"] "does foo") (TestTree "Foo" [TestTree "Bar" [TestSpec (TestInfo True "does bar"), TestTree "Baz" []]])
          `shouldBe` TestTree
            "Foo"
            [ TestTree
                "Bar"
                [ TestSpec (TestInfo True "does bar"),
                  TestTree
                    "Baz"
                    [TestSpec (TestInfo True "does foo")]
                ]
            ]
    describe "fromQuickLog" $ do
      it "is identical to insert for a single item" $
        fromQuickLog [QuickLogCasePass ["QuickSpec", "Foo"] "does foo"]
          `shouldBe` TestTree
            "QuickSpec"
            [ TestTree "Foo" [TestSpec (TestInfo True "does foo")]
            ]
      it "turns 'QuickSpec.X' into 'QuickSpec, X" $
        fromQuickLog [QuickLogCasePass ["QuickSpec.Foo"] "does foo"]
          `shouldBe` TestTree
            "QuickSpec"
            [ TestTree "Foo" [TestSpec (TestInfo True "does foo")]
            ]
      it "adds multiple items" $
        fromQuickLog [QuickLogCasePass ["QuickSpec", "Foo"] "does foo", QuickLogCasePass ["QuickSpec", "Foo"] "does bar"]
          `shouldBe` TestTree
            "QuickSpec"
            [ TestTree
                "Foo"
                [ TestSpec (TestInfo True "does foo"),
                  TestSpec (TestInfo True "does bar")
                ]
            ]


test/Hurtig/QuickParserSpec.hs
module Hurtig.QuickParserSpec (spec) where

import Data.Either
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import Hurtig.QuickLog
import Hurtig.QuickParser
import Test.Hspec
import Text.Megaparsec

spec :: Spec
spec = do
  describe "Hurtig" $ do
    describe "QuickParser" $ do
      describe "timeStamp" $ do
        it "should parse time stamp of the format YYYY-MM-DD HH:MM:SS.MS" $ do
          case runParser timeStamp "" "2022-12-17 23:08:03.162" of
            Right res -> res `shouldBe` UTCTime 2022 12 17 23 8 3 162
            Left x -> expectationFailure (show x)
      describe "quickLogOutput" $ do
        it "parses the start of a test suite" $ do
          case runParser quickLogOutput "" "Test Suite 'All Tests' started at 2022-12-17 23:08:03.162" of
            Right res -> res `shouldBe` QuickLogSuiteStart "All Tests" (UTCTime 2022 12 17 23 8 3 162)
            Left x -> expectationFailure $ show x
        it "parses the successful passing of a test suite" $ do
          case runParser quickLogOutput "" "Test Suite 'All Tests' passed at 2022-12-17 23:08:03.162" of
            Right res -> res `shouldBe` QuickLogSuitePass "All Tests" (UTCTime 2022 12 17 23 8 3 162)
            Left x -> expectationFailure $ show x
        it "parses the start of a test case" $ do
          case runParser quickLogOutput "" "Test Case 'QuickSpec.SwiftFoo, Util, getInput, reads data from file' started at 2022-12-17 23:08:03.163" of
            Right res -> res `shouldBe` QuickLogCaseStart ["QuickSpec.SwiftFoo", "Util", "getInput", "reads data from file"] (UTCTime 2022 12 17 23 8 3 163)
            Left x -> expectationFailure $ show x
        it "parses the passing of a test case" $ do
          case runParser quickLogOutput "" "Test Case 'QuickSpec.SwiftFoo, Util, getInput, reads data from file' passed (0.001 seconds)" of
            Right res -> res `shouldBe` QuickLogCasePass ["QuickSpec.SwiftFoo", "Util", "getInput"] "reads data from file"
            Left x -> expectationFailure $ show x
        it "parses Building Info output as QuickLogInfo" $ do
          case runParser quickLogOutput "" "Building for debugging..." of
            Right res -> res `shouldBe` QuickLogInfo "Building for debugging..."
            Left x -> expectationFailure $ show x
        it "parses Linking Info output as QuickLogInfo" $ do
          case runParser quickLogOutput "" "[13/13] Linking SwiftFooPackageTest.xctest" of
            Right res -> res `shouldBe` QuickLogInfo "[13/13] Linking SwiftFooPackageTest.xctest"
            Left x -> expectationFailure $ show x
        it "parses Build completion Info output as QuickLogInfo" $ do
          case runParser quickLogOutput "" "Build complete! (1.78s)" of
            Right res -> res `shouldBe` QuickLogInfo "Build complete"
            Left x -> expectationFailure $ show x
        it "parses Test Execution Info output as QuickLogInfo" $ do
          case runParser quickLogOutput "" "         Executed 4 tests, with 0 failures (0 unexpected) in 0.001 (0.001) seconds" of
            Right res -> res `shouldBe` QuickLogInfo "Executed 4 tests with 0 failures"
            Left x -> expectationFailure $ show x
        it "parses an expectation failure" $ do
          case runParser quickLogOutput "" "/Tests/QuickTests/FooTest.swift:24: error: QuickSpec.SwiftFoo, Util, composition, composes two functions : failed - expected to equal <7>, got <6>" of
            Right res -> res `shouldBe` QuickLogExpectationFailure ["QuickSpec.SwiftFoo", "Util", "composition", "composes two functions"] "expected to equal <7>, got <6>"
            Left x -> expectationFailure $ show x
        it "parses the full output of a successful run" $ do
          successOutput <- T.lines <$> TIO.readFile "./smoke-tests/quick-simple-success.in"
          let res = traverse (runParser quickLogOutput "") successOutput
          isRight res `shouldBe` True
        it "parses the full output of a failed run" $ do
          failedOutput <- T.lines <$> TIO.readFile "./smoke-tests/quick-failed-test.in"
          let res = traverse (runParser quickLogOutput "") (filter (/= "") failedOutput)
          case res of
            Right _ -> pure ()
            Left err -> expectationFailure $ show err


src/Hurtig/TestTree.hs
{-# LANGUAGE DeriveTraversable #-}

module Hurtig.TestTree where

import qualified Data.List as L
import Hurtig.QuickLog

data TestTree a
  = TestTree String [TestTree a]
  | TestSpec a
  deriving (Eq, Show, Functor, Foldable, Traversable)

data TestInfo = TestInfo
  { passed :: Bool,
    desc :: String
  }
  deriving (Eq, Show)

type QuickTestTree = TestTree TestInfo

insert :: QuickLog -> QuickTestTree -> QuickTestTree
insert (QuickLogCasePass _ _) (TestSpec x) = TestSpec x
insert (QuickLogCasePass path exp) tree@(TestTree name [])
  | length path == 1 && head path == name = TestTree name [TestSpec $ TestInfo True exp]
  | length path > 1 && head path == name = TestTree name [insert (QuickLogCasePass (tail path) exp) (TestTree (head $ tail path) [])]
  | otherwise = tree
insert (QuickLogCasePass path exp) tree@(TestTree name ts)
  | length path == 1 && head path == name = TestTree name (TestSpec (TestInfo True exp) : ts)
  | length path > 1 && head path == name = TestTree name (insert (QuickLogCasePass (tail path) exp) <$> ts)
  | otherwise = tree
insert _ tree = tree

fromQuickLog :: [QuickLog] -> QuickTestTree
fromQuickLog qs = foldr insert (TestTree "QuickSpec" []) (sanitizeTestCase <$> filter isTestCase qs)
  where
    isTestCase :: QuickLog -> Bool
    isTestCase (QuickLogCasePass _ _) = True
    isTestCase _ = False
    sanitizeTestCase :: QuickLog -> QuickLog
    sanitizeTestCase q@(QuickLogCasePass [] _) = q
    sanitizeTestCase q@(QuickLogCasePass (p : ps) _)
      | "QuickSpec." `L.isPrefixOf` p = q {testCasePath = "QuickSpec" : drop 10 p : ps}
      | otherwise = q
    sanitizeTestCase x = x


src/Hurtig/QuickParser.hs
module Hurtig.QuickParser where

import qualified Data.Text as T
import Data.Void
import Hurtig.QuickLog
import Text.Megaparsec
import Text.Megaparsec.Char
import Text.Megaparsec.Char.Lexer as L

type Parser = Parsec Void T.Text

singleQuotedString :: Parser String
singleQuotedString = char '\'' *> manyTill L.charLiteral (char '\'')

timeStamp :: Parser UTCTime
timeStamp = do
  year <- read <$> count 4 digitChar <* char '-'
  month <- read <$> count 2 digitChar <* char '-'
  day <- read <$> count 2 digitChar <* char ' '
  hour <- read <$> count 2 digitChar <* char ':'
  minute <- read <$> count 2 digitChar <* char ':'
  second <- read <$> count 2 digitChar <* char '.'
  ms <- read <$> count 3 digitChar
  return $ UTCTime year month day hour minute second ms

suiteParser :: (String -> UTCTime -> a) -> T.Text -> Parser a
suiteParser f str = do
  _ <- string "Test Suite" <* space1
  name <- singleQuotedString <* space1
  _ <- string str <* space1
  f name <$> timeStamp

commaSep :: Parser [String]
commaSep = sepBy (some $ alphaNumChar <|> char '.' <|> char ' ') (char ',' <* space1)

caseStartParser :: Parser QuickLog
caseStartParser = do
  _ <- string "Test Case" <* space1
  _ <- char '\''
  str <- commaSep
  _ <- char '\'' <* space1
  _ <- string "started at" <* space1
  QuickLogCaseStart str <$> timeStamp

casePassParser :: Parser QuickLog
casePassParser = do
  _ <- string "Test Case" <* space1
  _ <- char '\''
  str <- commaSep
  _ <- char '\'' <* space1
  _ <- string "passed" <* space1
  let testPath = init str
  let testDesc = last str
  return $ QuickLogCasePass testPath testDesc

caseFailParser :: Parser QuickLog
caseFailParser = do
  _ <- string "Test Case" <* space1
  _ <- char '\''
  str <- commaSep
  _ <- char '\'' <* space1
  _ <- string "failed" <* space1
  return $ QuickLogCaseFail str

expectationFailParser :: Parser QuickLog
expectationFailParser = do
  _ <- pathAndLineNo
  space1
  _ <- string "error:" <* space1
  casePath <- commaSep
  _ <- string ": failed - "
  -- TODO: I am only dropping whitespace here because i failed to adjust
  --       commaSep to not include trailing whitespace. I should figure
  --       this out ;/
  QuickLogExpectationFailure (dropWhiteSpace <$> casePath) . T.unpack <$> takeRest
  where
    pathAndLineNo = char '/' *> manyTill anySingle (char ':') <* decimal <* char ':'
    dropWhiteSpace = reverse . dropWhile (== ' ') . reverse

quickLogOutput :: Parser QuickLog
quickLogOutput = try suiteStart <|> try suitePass <|> try suiteFail <|> try caseStartParser <|> try casePassParser <|> try caseFailParser <|> try expectationFailParser <|> miscInfo
  where
    suiteStart = suiteParser QuickLogSuiteStart "started at"
    suitePass = suiteParser QuickLogSuitePass "passed at"
    suiteFail = suiteParser QuickLogSuiteFail "failed at"

parseQuickOutput :: String -> Either (ParseErrorBundle T.Text Void) [QuickLog]
parseQuickOutput str =
  let input = T.lines . T.pack $ str
   in traverse (runParser quickLogOutput "") input

miscInfo :: Parser QuickLog
miscInfo = try buildComplete <|> buildInfo <|> linkInfo <|> execInfo <|> emptyString
  where
    emptyString = do
      eof
      return $ QuickLogInfo ""
    buildInfo = do
      _ <- string "Building for"
      text <- some charLiteral
      return $ QuickLogInfo $ "Building for" ++ text

    linkInfo = do
      _ <- char '['
      x <- decimal
      _ <- char '/'
      y <- decimal
      _ <- char ']' <* space1
      _ <- string "Linking" <* space1
      name <- some (alphaNumChar <|> char '.')
      return $ QuickLogInfo $ "[" ++ show x ++ "/" ++ show y ++ "] Linking " ++ name

    buildComplete = do
      _ <- string "Build complete!"
      return $ QuickLogInfo "Build complete"

    execInfo = do
      _ <- some space1
      _ <- string "Executed" <* space1
      numTests <- decimal <* space1
      _ <- string "tests, with" <* space1
      numFail <- decimal <* space1
      _ <- string "failures" <|> string "failure"
      return $ QuickLogInfo $ "Executed " ++ show numTests ++ " tests with " ++ show numFail ++ " failures"


src/Hurtig/QuickLog.hs
module Hurtig.QuickLog where

import qualified Data.List as L
import qualified Data.Text as T
import Prettyprinter
import Prettyprinter.Render.String

data UTCTime = UTCTime
  { year :: Int,
    month :: Int,
    day :: Int,
    hour :: Int,
    minute :: Int,
    second :: Int,
    ms :: Int
  }
  deriving (Show, Eq)

data QuickLog
  = -- | '''Test Suite 'All tests' started at 2022-12-17 23:08:03.162'''
    QuickLogSuiteStart
      { suiteName :: String,
        suiteStart :: UTCTime
      }
  | -- | ''' Test Suite 'All tests' passed at 2022-12-17 23:08:03.164'''
    QuickLogSuitePass
      { suiteName :: String,
        suiteFinished :: UTCTime
      }
  | -- | ''' Test Suite 'All tests' failed at 2022-12-17 23:08:03.164'''
    QuickLogSuiteFail
      { suiteName :: String,
        suiteFinished :: UTCTime
      }
  | -- | '''Test Case 'QuickSpec.SwiftFoo, Util, getInput, reads data from file' started at 2022-12-17 23:08:03.163'''
    QuickLogCaseStart
      { testCasePath :: [String],
        testCaseStart :: UTCTime
      }
  | -- | '''Test Case 'QuickSpec.SwiftFoo, Util, getInput, reads data from file' passed (0.001 seconds)'''
    QuickLogCasePass
      { testCasePath :: [String],
        desc :: String
      }
  | -- | '''Test Case 'QuickSpec.SwiftFoo, Util, getInput, reads data from file' passed (0.001 seconds)'''
    QuickLogCaseFail
      {testCasePath :: [String]}
  | -- | /Tests/QuickTests/FooTest.swift:24: error: QuickSpec.SwiftFoo, Util, composition, composes two functions : failed - expected to equal <7>, got <6>
    QuickLogExpectationFailure
      { testCasePath :: [String],
        errorInfo :: String
      }
  | -- | Any other output
    QuickLogInfo {info :: String}
  deriving (Eq, Show)

prettyUTCTime :: UTCTime -> Doc ann
prettyUTCTime (UTCTime y m d h min s ms) =
  mconcat
    [ pretty y,
      "-",
      pretty m,
      "-",
      pretty d,
      " ",
      pretty h,
      ":",
      pretty min,
      ":",
      pretty s,
      ".",
      pretty ms
    ]

printQuickLog :: QuickLog -> String
printQuickLog = renderString . layoutPretty defaultLayoutOptions . prettyQuickLog

prettyQuickLog :: QuickLog -> Doc ann
prettyQuickLog ql = case ql of
  QuickLogSuiteStart name time ->
    "Test Suite '" <> pretty name <> "' started at " <> prettyUTCTime time
  QuickLogSuitePass name time ->
    "Test Suite '" <> pretty name <> "' passed at " <> prettyUTCTime time
  QuickLogSuiteFail name time ->
    "Test Suite '" <> pretty name <> "' failed at " <> prettyUTCTime time
  QuickLogCaseStart names time ->
    let p = L.intersperse ", " $ init names
        path = hcat (map (pretty . T.pack) p)
        testDesc = last names
     in "Test Case '" <> path <> ", " <> pretty testDesc <> "' started at " <> prettyUTCTime time
  QuickLogCasePass casePath desc ->
    "Test Case '" <> hcat (map (pretty . T.pack) (casePath ++ [desc])) <> "' passed"
  QuickLogCaseFail names ->
    "Test Case '" <> hcat (map (pretty . T.pack) names) <> "' failed"
  QuickLogExpectationFailure p info ->
    "error: " <> hcat (map (pretty . T.pack) p) <> " failed - " <> pretty info
  QuickLogInfo s ->
    pretty s
